File: backend/app/services/invite_service.py
"""邀请码服务:生成、列表、失效、续期。"""

from __future__ import annotations

import secrets
from datetime import datetime, timedelta

from sqlalchemy.orm import Session

from app.models.invite_codes import InviteCode
from app.models.users import User
from app.schemas.common import AppError, ErrorCode
from app.schemas.invite import InviteActionOut, InviteOut
from app.utils.time import utcnow

INVITE_CODE_LENGTH = 8
_INVITE_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"


def generate_invite_code(db: Session) -> str:
    for _ in range(20):
        code = "".join(secrets.choice(_INVITE_ALPHABET) for _ in range(INVITE_CODE_LENGTH))
        existing = db.query(InviteCode).filter(InviteCode.code == code).first()
        if not existing:
            return code
    raise AppError(ErrorCode.INTERNAL, "Failed to generate unique invite code", 500)


def _compute_expires_at(duration_days: int | None) -> datetime | None:
    if duration_days is None:
        return None
    return utcnow() + timedelta(days=duration_days)


def _sync_expired(db: Session, creator_id: int) -> None:
    now = utcnow()
    db.query(InviteCode).filter(
        InviteCode.creator_id == creator_id,
        InviteCode.status == "active",
        InviteCode.expires_at.isnot(None),
        InviteCode.expires_at <= now,
    ).update({"status": "expired"}, synchronize_session="fetch")
    db.commit()


def create_invite(db: Session, user: User, data: dict) -> dict:
    if not user.can_invite:
        raise AppError(ErrorCode.INVITE_DISABLED, "Invite permission disabled", 403)
    _sync_expired(db, user.id)
    active = (
        db.query(InviteCode)
        .filter(InviteCode.creator_id == user.id, InviteCode.status == "active")
        .first()
    )
    if active:
        raise AppError(
            ErrorCode.ACTIVE_INVITE_EXISTS,
            "Active invite already exists, revoke or renew first",
            400,
        )
    code = generate_invite_code(db)
    expires_at = _compute_expires_at(data.get("duration_days"))
    invite = InviteCode(
        creator_id=user.id,
        code=code,
        status="active",
        expires_at=expires_at,
    )
    db.add(invite)
    db.commit()
    db.refresh(invite)
    return InviteOut(
        id=invite.id,
        code=invite.code,
        status=invite.status,
        expires_at=invite.expires_at,
        created_at=invite.created_at,
        used_by_id=invite.used_by_id,
    ).model_dump(mode="json")


def list_invites(db: Session, user: User) -> list[dict]:
    _sync_expired(db, user.id)
    rows = (
        db.query(InviteCode)
        .filter(InviteCode.creator_id == user.id)
        .order_by(InviteCode.created_at.desc())
        .all()
    )
    return [
        InviteOut(
            id=r.id,
            code=r.code,
            status=r.status,
            expires_at=r.expires_at,
            created_at=r.created_at,
            used_by_id=r.used_by_id,
        ).model_dump(mode="json")
        for r in rows
    ]


def revoke_invite(db: Session, user: User, invite_id: int) -> dict:
    invite = db.query(InviteCode).filter(InviteCode.id == invite_id).first()
    if not invite or invite.creator_id != user.id:
        raise AppError(ErrorCode.INVITE_NOT_FOUND, "Invite not found", 404)
    if invite.status == "used":
        raise AppError(ErrorCode.INVITE_ALREADY_USED, "Used invite cannot be revoked", 400)
    if invite.status in ("revoked", "expired"):
        return {"message": "Invite already inactive"}
    invite.status = "revoked"
    db.commit()
    return InviteActionOut(message="Invite revoked").model_dump(mode="json")


def renew_invite(db: Session, user: User, data: dict) -> dict:
    if not user.can_invite:
        raise AppError(ErrorCode.INVITE_DISABLED, "Invite permission disabled", 403)
    _sync_expired(db, user.id)
    active = (
        db.query(InviteCode)
        .filter(InviteCode.creator_id == user.id, InviteCode.status == "active")
        .first()
    )
    if not active:
        raise AppError(
            ErrorCode.INVITE_NOT_FOUND,
            "No active invite to renew, create one first",
            404,
        )
    duration_days = data.get("duration_days")
    if duration_days is None:
        active.expires_at = None
    else:
        now = utcnow()
        base = active.expires_at if active.expires_at and active.expires_at > now else now
        active.expires_at = base + timedelta(days=duration_days)
    db.commit()
    db.refresh(active)
    return InviteOut(
        id=active.id,
        code=active.code,
        status=active.status,
        expires_at=active.expires_at,
        created_at=active.created_at,
        used_by_id=active.used_by_id,
    ).model_dump(mode="json")

File: backend/internal/service/invite_service.go
package service

import (
	"context"
	"crypto/rand"
	"math/big"
	"time"

	"backend/internal/model"
	"backend/internal/repository"
)

const inviteAlphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"

type InviteService struct {
	repos *repository.Repositories
}

func NewInviteService(repos *repository.Repositories) *InviteService {
	return &InviteService{repos: repos}
}

func (s *InviteService) generateCode(ctx context.Context) (string, error) {
	for i := 0; i < 20; i++ {
		b := make([]byte, 8)
		for j := range b {
			idx, _ := rand.Int(rand.Reader, big.NewInt(int64(len(inviteAlphabet))))
			b[j] = inviteAlphabet[idx.Int64()]
		}
		code := string(b)
		existing, err := s.repos.Invite.GetByCode(ctx, code)
		if err != nil {
			return "", err
		}
		if existing == nil {
			return code, nil
		}
	}
	return "", model.NewAppError(model.ErrInternal, "生成唯一邀请码失败，请重试", 500)
}

func (s *InviteService) CreateInvite(ctx context.Context, user *model.User, durationDays *int) (*model.InviteOut, error) {
	if !user.CanInvite {
		return nil, model.NewAppError(model.ErrInviteDisabled, "邀请权限已被关闭", 403)
	}

	_ = s.repos.Invite.SyncExpired(ctx, user.ID)

	active, err := s.repos.Invite.GetActiveByCreator(ctx, user.ID)
	if err != nil {
		return nil, err
	}
	if active != nil {
		return nil, model.NewAppError(model.ErrActiveInviteExists, "已有有效邀请码，请先失效或续期", 400)
	}

	code, err := s.generateCode(ctx)
	if err != nil {
		return nil, err
	}

	var expiresAt *time.Time
	if durationDays != nil {
		t := time.Now().UTC().Add(time.Duration(*durationDays) * 24 * time.Hour)
		expiresAt = &t
	}

	inv := &model.InviteCode{
		Code:      code,
		CreatorID: user.ID,
		Status:    "active",
		ExpiresAt: expiresAt,
	}

	if err := s.repos.Invite.Create(ctx, inv); err != nil {
		return nil, model.NewAppError(model.ErrInternal, "创建邀请码失败", 500)
	}

	return &model.InviteOut{
		ID:        inv.ID,
		Code:      inv.Code,
		Status:    inv.Status,
		ExpiresAt: inv.ExpiresAt,
		CreatedAt: inv.CreatedAt,
		UsedByID:  inv.UsedByID,
	}, nil
}

func (s *InviteService) ListInvites(ctx context.Context, user *model.User) ([]model.InviteOut, error) {
	_ = s.repos.Invite.SyncExpired(ctx, user.ID)

	invites, err := s.repos.Invite.ListByCreator(ctx, user.ID)
	if err != nil {
		return nil, err
	}

	var result []model.InviteOut
	for _, inv := range invites {
		result = append(result, model.InviteOut{
			ID:        inv.ID,
			Code:      inv.Code,
			Status:    inv.Status,
			ExpiresAt: inv.ExpiresAt,
			CreatedAt: inv.CreatedAt,
			UsedByID:  inv.UsedByID,
		})
	}
	if result == nil {
		result = []model.InviteOut{}
	}
	return result, nil
}

func (s *InviteService) RevokeInvite(ctx context.Context, user *model.User, inviteID int) (*model.InviteActionOut, error) {
	inv, err := s.repos.Invite.GetByID(ctx, inviteID)
	if err != nil {
		return nil, err
	}
	if inv == nil || inv.CreatorID != user.ID {
		return nil, model.NewAppError(model.ErrInviteNotFound, "邀请码不存在", 404)
	}

	if inv.Status == "used" {
		return nil, model.NewAppError(model.ErrInviteAlreadyUsed, "已使用的邀请码无法失效", 400)
	}

	if inv.Status == "revoked" || inv.Status == "expired" {
		return &model.InviteActionOut{Message: "邀请码已失效"}, nil
	}

	if err := s.repos.Invite.Revoke(ctx, inviteID); err != nil {
		return nil, model.NewAppError(model.ErrInternal, "失效邀请码失败", 500)
	}

	return &model.InviteActionOut{Message: "邀请码已失效"}, nil
}

func (s *InviteService) RenewInvite(ctx context.Context, user *model.User, durationDays *int) (*model.InviteOut, error) {
	if !user.CanInvite {
		return nil, model.NewAppError(model.ErrInviteDisabled, "邀请权限已被关闭", 403)
	}

	_ = s.repos.Invite.SyncExpired(ctx, user.ID)

	active, err := s.repos.Invite.GetActiveByCreator(ctx, user.ID)
	if err != nil {
		return nil, err
	}
	if active == nil {
		return nil, model.NewAppError(model.ErrInviteNotFound, "没有可续期的邀请码，请先生成一个", 404)
	}

	var newExpiresAt *time.Time
	if durationDays != nil {
		now := time.Now().UTC()
		base := now
		if active.ExpiresAt != nil && active.ExpiresAt.After(now) {
			base = *active.ExpiresAt
		}
		t := base.Add(time.Duration(*durationDays) * 24 * time.Hour)
		newExpiresAt = &t
	}

	if err := s.repos.Invite.Renew(ctx, active.ID, newExpiresAt); err != nil {
		return nil, model.NewAppError(model.ErrInternal, "续期邀请码失败", 500)
	}

	active.ExpiresAt = newExpiresAt
	return &model.InviteOut{
		ID:        active.ID,
		Code:      active.Code,
		Status:    active.Status,
		ExpiresAt: active.ExpiresAt,
		CreatedAt: active.CreatedAt,
		UsedByID:  active.UsedByID,
	}, nil
}
